feat: cloud.Topic readme docs (#2667)
winglang/wing: libs/wingsdk/src/cloud/topic.md

---
title: cloud.Topic 
id: topic
description: A built-in resource for publishing messages to subscribers.
keywords: [Wing reference, Wing language, language, Wing sdk, Wing programming language, topics]
---

The `cloud.Topic` class represents a subject of data that is open for subscription.

Topics are a staple of event-driven architectures, especially those that rely on pub-sub messaging to decouple producers of data and the consumers of said data. 

## Usage

### Creating a topic

```js
bring cloud;

let topic = new cloud.Topic();
```

> ### Using a topic

Reviewer: ```suggestion
### Subscribing to a topic
```

```js
bring cloud;

let topic = new cloud.Topic();

topic.onMessage(inflight (message: str) => {
  log("Topic published message: ${message}");
});
```

> ### Using a topic inflight
> The inflight api for a topic allows for publishing messages to the topic.

Reviewer: (Thinking out loud) Maybe using old-school OO terms like "method" will reinforce the idea that these are just ordinary classes in the language
```suggestion
### Publishing to a topic

The inflight method `publish` sends a message to all of the topic's subscribers.
```

```js
bring cloud;

let topic = new cloud.Topic();

> inflight() => {

Reviewer: ```suggestion
inflight () => {
```

  topic.publish("Hello World!");
};
```

### Simple pub-sub example

Here is an example of combining the preflight and inflight apis for a topic and creating an adorable
simple pub-sub application.

```js
bring cloud;

// First we create a topic
let topic = new cloud.Topic();

// Then we define a consumer inflight handler
let consumerHandler = inflight(message: str) => {
  log("Doing some work with message: ${message}");
};

> // Now we can use the preflight api of topic to register the consumer handler

Reviewer: ```suggestion
// Now we can use a preflight method of topic to register the consumer handler
```

// to be invoked when a message is published to the topic.
topic.onMessage(consumerHandler);

// Then we define the producer inflight handler

> let publisherHandler = inflight() => {

Reviewer: ```suggestion
let publisherHandler = inflight () => {
```
